staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go

```diff
@@ -0,0 +1,220 @@
+	maxBatchWait time.Duration
+}
+
+// NewBackend returns an audit backend at sends events over HTTP to an external service.
+// The mode indicates the caching behavior of the webhook. Either blocking (ModeBlocking)
+// or buffered with batch POSTs (ModeBatch).
+func NewBackend(kubeConfigFile string, mode string) (*AuditBackend, error) {
+	// NOTE: Copied from other webhook implementations
+	//
+	// TODO: Get these out of the init() statements of other webhok implementations?
+	// Can we make these passable to NewGenericWebhook?
+	groupFactoryRegistry := make(announced.APIGroupFactoryRegistry)
+	groupVersions := []schema.GroupVersion{auditv1alpha1.SchemeGroupVersion}
+	registry := registered.NewOrDie("")
+
+	registry.RegisterVersions(groupVersions)
+	if err := registry.EnableVersions(groupVersions...); err != nil {
+		return nil, fmt.Errorf("failed to enable version %v", groupVersions)
+	}
+	install.Install(groupFactoryRegistry, registry, audit.Scheme)
+
+	w, err := webhook.NewGenericWebhook(registry, audit.Codecs, kubeConfigFile, groupVersions, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	b := &AuditBackend{w: w}
+
+	switch mode {
+	case ModeBatch:
+		// enable batching.
+		b.buffer = make(chan *auditinternal.Event, defaultBatchBufferSize)
+		b.maxBatchSize = defaultBatchMaxSize
+		b.maxBatchWait = defaultBatchMaxWait
+	case ModeBlocking:
+	default:
+		return nil, fmt.Errorf("webhook mode %q is not in list of known modes (%s)",
+			mode, strings.Join(AllowedModes, ","))
+	}
+
+	return b, nil
+}
+
+// Run starts the webhook and shuts down once stopCh closes. It MUST be called before
+// using the backend. Run is non-blocking.
+func (a *AuditBackend) Run(stopCh <-chan struct{}) error {
+	if a.buffer == nil {
+		// Not buffered, nothing to do.
+		return nil
+	}
+
+	f := func() {
+		// Recover from any panics caused by this method so a panic in the
+		// goroutine can't bring down the main routine.
+		defer runtime.HandleCrash()
+
+		t := time.NewTicker(a.maxBatchWait)
+		defer t.Stop() // Release ticker resources
+
+		if err := a.sendBatchEvents(stopCh, t.C); err != nil {
+			glog.Errorf("failed to POST webhook events: %v", err)
+		}
+	}
+
+	go func() {
+		for {
+			f()
+		}
+	}()
+
+	return nil
+}
+
+// ProcessEvents attempts to POST a serialized audit event to an external service. The actual
+// payload contains an audit.EventList with a single element. The audit webhook may batch
+// multiple events in the future.
+func (a *AuditBackend) ProcessEvents(ev ...*auditinternal.Event) {
+	if a.buffer != nil {
+		a.queueEvents(ev...)
+		return
+	}
+
+	if err := a.sendEvents(ev...); err != nil {
+		glog.Errorf("failed to POST webhook events: %v", err)
+	}
+}
+
+// queueEvents adds the provided events to the internal buffer.
+func (a *AuditBackend) queueEvents(ev ...*auditinternal.Event) {
+	for _, e := range ev {
+		select {
+		case a.buffer <- e:
+		default:
+			glog.Errorf("events queue blocked, failed to send event %#v", e)
```

Comment:
I was planning to send a follow up PR to do this. Up to you whether you want to wait for that.